File: app/src/main/java/com/hana053/micropost/presentation/relateduserlist/RelatedUserListAdapter.java
package com.hana053.micropost.presentation.relateduserlist;

import android.support.annotation.Nullable;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.ViewGroup;

import com.hana053.micropost.databinding.UserItemBinding;
import com.hana053.micropost.domain.RelatedUser;
import com.hana053.micropost.presentation.core.components.avatar.AvatarViewListener;
import com.hana053.micropost.presentation.core.components.avatar.AvatarViewModel;
import com.hana053.micropost.presentation.core.components.followbtn.FollowBtnViewListener;
import com.hana053.micropost.presentation.core.components.followbtn.FollowBtnViewModel;
import com.hana053.micropost.presentation.core.services.AuthTokenService;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

import rx.Observable;

public class RelatedUserListAdapter extends RecyclerView.Adapter<RelatedUserListAdapter.ViewHolder> {

    private final List<RelatedUser> users = new ArrayList<>();
    private final AuthTokenService authTokenService;

    public RelatedUserListAdapter(AuthTokenService authTokenService) {
        this.authTokenService = authTokenService;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        return ViewHolder.newInstance(parent);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        final RelatedUser relatedUser = users.get(position);
        final boolean isMyself = authTokenService.isMyself(relatedUser);

        final RelatedUserItemViewModel viewModel = new RelatedUserItemViewModel(relatedUser, isMyself);
        holder.binding.setModel(viewModel);

        final FollowBtnViewModel followBtnViewModel = new FollowBtnViewModel(relatedUser.getId(), relatedUser.followedByMe);
        holder.binding.setFollowBtnModel(followBtnViewModel);
        holder.binding.setFollowBtnListener(new FollowBtnViewListener(followBtnViewModel));

        final AvatarViewModel avatarModel = new AvatarViewModel(relatedUser);
        holder.binding.setAvatarModel(avatarModel);
        holder.binding.setAvatarListener(new AvatarViewListener(avatarModel));
    }

    @Override
    public int getItemCount() {
        return users.size();
    }

    List<RelatedUser> getItems() {
        return new ArrayList<>(users);
    }

    static class ViewHolder extends RecyclerView.ViewHolder {
        private final UserItemBinding binding;

        ViewHolder(UserItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }

        private static ViewHolder newInstance(ViewGroup parent) {
            final LayoutInflater inflater = LayoutInflater.from(parent.getContext());
            final UserItemBinding binding = UserItemBinding.inflate(inflater, parent, false);
            return new ViewHolder(binding);
        }
    }

    @Nullable
    Long getLastItemId() {
        return Observable.from(users)
                .takeLast(1)
                .map(RelatedUser::getRelationshipId)
                .toBlocking()
                .singleOrDefault(null);
    }

    public boolean addAll(int location, Collection<RelatedUser> users) {
        if (this.users.addAll(location, users)) {
            notifyItemRangeInserted(location, users.size());
            return true;
        }
        return false;
    }

}

File: app/src/main/kotlin/com/hana053/micropost/pages/relateduserlist/RelatedUserListAdapter.kt
package com.hana053.micropost.pages.relateduserlist

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import com.hana053.micropost.R
import com.hana053.micropost.domain.RelatedUser
import com.hana053.micropost.domain.User
import com.hana053.micropost.shared.avatar.AvatarView
import com.hana053.micropost.shared.followbtn.FollowBtnView
import kotlinx.android.synthetic.main.item_related_users.view.*
import rx.subjects.PublishSubject


class RelatedUserListAdapter(
    private val users: MutableList<RelatedUser> = mutableListOf()
) : RecyclerView.Adapter<RelatedUserListAdapter.ViewHolder>() {

    val followBtnClicksSubject: PublishSubject<FollowBtnView> = PublishSubject.create()
    val avatarClicksSubject: PublishSubject<User> = PublishSubject.create()

    class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        val container: LinearLayout = view.container
        val avatar: ImageView = view.avatar
        val userName: TextView = view.userName
        val followBtn: Button = view.followBtn
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) =
        LayoutInflater.from(parent.context)
            .inflate(R.layout.item_related_users, parent, false)
            .let(::ViewHolder)

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val item = users[position]

        holder.apply {
            container.tag = item
            userName.text = item.name

            AvatarView(avatar).render(item.toUser())
            avatar.setOnClickListener {
                avatarClicksSubject.onNext(item.toUser())
            }

            FollowBtnView(followBtn).apply {
                render(item.toUser())
                followBtn.setOnClickListener {
                    followBtnClicksSubject.onNext(this)
                }
            }
        }
    }

    override fun getItemCount() = users.size

    fun getLastItemId() = users.map { it.relationshipId }.lastOrNull()

    fun addAll(location: Int, users: List<RelatedUser>): Boolean {
        if (this.users.addAll(location, users)) {
            notifyItemRangeInserted(location, users.size)
            return true
        }
        return false
    }

}